src/Bridge/Doctrine/MongoDbOdm/PropertyHelperTrait.php

```diff
@@ -0,0 +1,91 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Doctrine\MongoDbOdm;
+
+use ApiPlatform\Core\Exception\InvalidArgumentException;
+use Doctrine\Common\Persistence\Mapping\ClassMetadata;
+use Doctrine\ODM\MongoDB\Aggregation\Builder;
+use Doctrine\ODM\MongoDB\Mapping\ClassMetadata as MongoDbOdmClassMetadata;
+
+/**
+ * Helper trait regarding a property in a MongoDB document using the resource metadata.
+ *
+ * @experimental
+ *
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+trait PropertyHelperTrait
+{
+    /**
+     * Splits the given property into parts.
+     */
+    abstract protected function splitPropertyParts(string $property/*, string $resourceClass*/): array;
+
+    /**
+     * Gets class metadata for the given resource.
+     */
+    abstract protected function getClassMetadata(string $resourceClass): ClassMetadata;
+
+    /**
+     * Adds the necessary lookups for a nested property.
+     *
+     * @throws InvalidArgumentException If property is not nested
+     *
+     * @return array An array where the first element is the $alias of the lookup,
+     *               the second element is the $field name
+     *               the third element is the $associations array
+     */
+    protected function addLookupsForNestedProperty(string $property, Builder $aggregationBuilder, string $resourceClass): array
+    {
+        $propertyParts = $this->splitPropertyParts($property, $resourceClass);
+        $alias = '';
+
+        foreach ($propertyParts['associations'] as $association) {
+            $classMetadata = $this->getClassMetadata($resourceClass);
+
+            if (!$classMetadata instanceof MongoDbOdmClassMetadata) {
+                break;
+            }
+
+            if ($classMetadata->hasReference($association)) {
+                $propertyAlias = "${association}_lkup";
+                // previous_association_lkup.association
+                $localField = "$alias$association";
+                // previous_association_lkup.association_lkup
+                $alias .= $propertyAlias;
+                $referenceMapping = $classMetadata->getFieldMapping($association);
+
+                $aggregationBuilder->lookup($classMetadata->getAssociationTargetClass($association))
```

Comment:
ODM does this for you, throwing a helpful exception if you're using a reference that can't be looked up:
```
$aggregationBuilder
    ->lookup($association)
        ->alias($alias)
```

Your code does not handle different reference storage strategies (`id`, `ref`) appropriately.

Reply:
I had to do like this to manage to have nested lookups. It's because there is a garding `if` in the `Lookup` stage code using `$this->class` (https://github.com/doctrine/mongodb-odm/blob/55e1b9d87f6d60042878c368262d072931737b53/lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Lookup.php#L154-L156).
I don't understand why you say it does not handle the different strategies. It's nearly the same code as in the `Lookup` class.